set.mm

```diff
@@ -751469,4 +751469,88 @@ with simplest antecedents (i.e., in the corresponding ordering of binary
+    cu3addd.2 $e |- ( ph -> B e. CC ) $.
+    cu3addd.3 $e |- ( ph -> C e. CC ) $.
+    $( Cube of sum of three numbers.  (Contributed by Igor Ieskov,
+       14-Dec-2023.) $)
+    cu3addd $p |- ( ph -> ( ( ( A + B ) + C ) ^ 3 ) =
+                ( ( ( ( ( A ^ 3 ) + ( 3 x. ( ( A ^ 2 ) x. B ) ) )
+                + ( ( 3 x. ( A x. ( B ^ 2 ) ) ) + ( B ^ 3 ) ) )
+                + ( ( ( 3 x. ( ( A ^ 2 ) x. C ) )
+                + ( ( ( 3 x. 2 ) x. ( A x. B ) ) x. C ) )
+                + ( 3 x. ( ( B ^ 2 ) x. C ) ) ) )
+                + ( ( ( 3 x. ( A x. ( C ^ 2 ) ) )
+                + ( 3 x. ( B x. ( C ^ 2 ) ) ) ) + ( C ^ 3 ) ) ) ) $=
+      ( co wcel wceq addcld binom3 a1i oveq1d eqtrd oveq2d sqcld mulcld adddird
+      adddid caddc c3 cexp c2 cmul cc wa jca wi mpd syl2anc binom2d 3cn mulassd
+      2cnd eqcomd ) ABCUAZHZDUQHUBZUCZHZBUSUTHUSBUDZUTHZCUEZHVDHUQHUSBCVBUTHZVD
+      HVDHCUSUTHUQHUQHZUSVCDVDHZVDHZUSVBVDHBCVDHZVDHZDVDHZUQHZUSVEDVDHZVDHZUQHZ
+      UQHZUSBDVBUTHZVDHZCVQVDHZUQHZVDHZDUSUTHZUQHZUQHZVPUSVRVDHUSVSVDHUQHZWBUQH
+      ZUQHAVAVPUSURVQVDHZVDHZWBUQHZUQHZWDAVAVFVHUSVBVIVDHZDVDHZVDHZUQHZVNUQHZUQ
+      HZWIUQHZWJAVAVFUSVGWLUQHZVDHZVNUQHZUQHZWIUQHZWQAVAVFUSWRVMUQHZVDHZUQHZWIU
+      QHZXBAVAVFUSVCWKUQHZDVDHZVMUQHZVDHZUQHZWIUQHZXFAVAVFUSXGVEUQHZDVDHZVDHZUQ
+      HZWIUQHZXLAVAVFUSURVBUTHZDVDHZVDHZUQHZWIUQHZXQAVAURUSUTHZXTUQHZWIUQHZYBAU
+      RUFZIZDYFIZUGZVAYEJZAYGYHABCEFKGUHYIYJUIAURDLMUJAYDYAWIUQAYCVFXTUQABYFICY
+      FIYCVFJEFBCLUKNNOAYAXPWIUQAXTXOVFUQAXSXNUSVDAXRXMDVDABCEFULNPPNOAXPXKWIUQ
+      AXOXJVFUQAXNXIUSVDAXGVEDAVCWKABEQZAVBVIAUOZABCEFRZRZKACFQZGSPPNOAXKXEWIUQ
+      AXJXDVFUQAXIXCUSVDAXHWRVMUQAVCWKDYKYNGSNPPNOAXEXAWIUQAXDWTVFUQAUSWRVMUSYF
+      IAUMMZAVGWLAVCDYKGRZAWKDYNGRZKAVEDYOGRTPNOAXAWPWIUQAWTWOVFUQAWSWNVNUQAUSV
+      GWLYPYQYRTNPNOAWPVPWIUQAVPWPAVOWOVFUQAVLWNVNUQAVKWMVHUQAVKUSWKVDHZDVDHWMA
+      VJYSDVDAUSVBVIYPYLYMUNNAUSWKDYPYNGUNOPNPUPNOAWIWCVPUQAWHWAWBUQAWGVTUSVDAB
+      CVQEFADGQZSPNPOAWCWFVPUQAWAWEWBUQAUSVRVSYPABVQEYTRACVQFYTRTNPO $.
+  $}
+
+  ${
+    cu3addi.1 $e |- A e. CC $.
+    cu3addi.2 $e |- B e. CC $.
+    cu3addi.3 $e |- C e. CC $.
+    $( Cube of sum of three numbers.  (Contributed by Igor Ieskov,
+       14-Dec-2023.) $)
+    $( Cube of sum of three numbers.  (Contributed by Igor Ieskov,
+       14-Dec-2023.) $)
+    cu3addi $p |- ( ( ( A + B ) + C ) ^ 3 ) =
+    ( ( ( ( ( A ^ 3 ) + ( 3 x. ( ( A ^ 2 ) x. B ) ) )
+    + ( ( 3 x. ( A x. ( B ^ 2 ) ) ) + ( B ^ 3 ) ) )
+    + ( ( ( 3 x. ( ( A ^ 2 ) x. C ) )
+    + ( ( ( 3 x. 2 ) x. ( A x. B ) ) x. C ) )
+    + ( 3 x. ( ( B ^ 2 ) x. C ) ) ) )
+    + ( ( ( 3 x. ( A x. ( C ^ 2 ) ) )
+    + ( 3 x. ( B x. ( C ^ 2 ) ) ) ) + ( C ^ 3 ) ) )
+    $=
+      ( caddc co c3 cexp c2 cmul wceq wn cc wcel a1i cu3addd pm2.18i ) ABGZHCTH
+      IZJZHAUAUBHUAAKZUBHZBLZHUEHTHUAABUCUBHZUEHUEHBUAUBHTHTHUAUDCUEHUEHUAUCUEH
+      ABUEHUEHCUEHTHUAUFCUEHUEHTHTHUAACUCUBHZUEHUEHUABUGUEHUEHTHCUAUBHTHTHMZUHN
+      ZABCAOZPUIDQBUJPUIEQCUJPUIFQRS $.
+  $}
```

Comment:
Interesting, that ~pm2.18i can be used instead of ~mptru (and therefore to avoid the usage ov "T.").

Reply:
This was found by mm-lamp. I even didn't pay attention to this.

Reply:
@avekens - I agree it's interesting, though in retrospect it's not surprising. Obviously this trick doesn't work in general in intuitionistic logic (iset.mm).

Reply:
The version with `mptru`` is shorter (199 bytes in SHOW PROOF/COMPRESSED compared with 206 bytes for the pm2.18i one). I guess the reason the difference is so small is that the compressor can reuse the long expression. Ah yes, and in SHOW PROOF/SIZE the difference is 227 characters versus 278).

Sorry for geeking out on this sort of minutia, @expln .  Us metamath folks are vulnerable to this sort of rabbit trap.

Reply:
Does this mean I have to change the proof? Or it is fine to use pm2.18i?

Reply:
Have to? Not on my account.

If you want to, feel free to make a follow up pull request, I guess I don't have a strong opinion on the matter although we do tend to prefer shorter proofs.

Reply:
> Have to?

Sorry, I don't know English very well, sometimes I use incorrect wording :)

As for the length of the proof. mm-lamp uses BFS to find a proof, where each tree node is an assertion application. If there are two possible proofs:

```
Proof1:
h1:: a short hypothesis
h2:: a short hypothesis
h3:: a short hypothesis
1:h1,h2,h3,h1,h2,h3,h1,h2,h3,h1,h2,h3,h1,h2,h3:asrt1: the goal statement
```

```
Proof2:
h1:: a short hypothesis
h2:: a short hypothesis
h3:: a short hypothesis
1:h1,h2:asrt2: a short intermediate statement
2:h3,1:asrt3: the goal statement
```

Then mm-lamp will find the first proof and stop searching, even though the first proof may be much longer then the second one. And I am not familiar enough with set.mm assertions and usual approaches, so I just submitted what the assistant found (after checking correctness of the proof by metamath.exe). The only assertion I remembered after creating these proofs is `addassi` because the assistant could not use it automatically somehow. But I will produce shorter proofs as I get more familiar with set.mm :)

Reply:
Just recently I looked at the difference between usage of pm2.18i and mptru and understood what's going on. I will submit a shorter proof soon.